kvdb/postgres/db.go

```diff
@@ -0,0 +1,233 @@
+// +build kvdb_postgres
+
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	_ "github.com/jackc/pgx/v4/stdlib"
+)
+
+const (
+	// hexTableId is the table name for top-level buckets data under a key
+	// that cannot be mapped to a distinct table.
+	hexTableId = "hex"
+
+	// sequenceTableId is the table name that holds top-level bucket
+	// sequence numbers.
+	sequenceTableId = "sequence"
+
+	// systemTablePrefixExtension is the prefix given to system tables.
+	systemTablePrefixExtension = "sys"
+)
+
+// KV stores a key/value pair.
+type KV struct {
+	key string
+	val string
+}
+
+// db holds a reference to the postgres connection connection.
+type db struct {
+	// cfg is the postgres connection config.
+	cfg *Config
+
+	// prefix is the table name prefix that is used to simulate namespaces.
+	// We don't use schemas because at least sqlite does not support that.
+	prefix string
+
+	// ctx is the overall context for the database driver.
+	//
+	// TODO: This is an anti-pattern that is in place until the kvdb
+	// interface supports a context.
+	ctx context.Context
+
+	// db is the underlying database connection instance.
+	db *sql.DB
+
+	// lock is the global write lock that ensures single writer.
+	lock sync.RWMutex
+
+	// hexTableName is the name of the table that contains the data for all
+	// top-level buckets that have keys that cannot be mapped to a distinct
+	// sql table.
+	hexTableName string
+
+	// sequenceTableName is the name of the table that holds the sequence
+	// numbers for for all top-level buckets that have keys that cannot be
+	// mapped to a distinct sql table.
+	sequenceTableName string
+}
+
+// Enforce db implements the walletdb.DB interface.
+var _ walletdb.DB = (*db)(nil)
+
+// newPostgresBackend returns a db object initialized with the passed backend
+// config. If postgres connection cannot be estabished, then returns error.
+func newPostgresBackend(ctx context.Context, config *Config, prefix string) (
+	*db, error) {
+
+	if prefix == "" {
+		return nil, errors.New("empty postgres prefix")
+	}
+
+	dbConn, err := sql.Open("pgx", config.Dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compose system table names.
+	hexTableName := fmt.Sprintf(
+		"%s%s_%s", prefix, systemTablePrefixExtension, hexTableId,
+	)
+
+	sequenceTableName := fmt.Sprintf(
+		"%s%s_%s", prefix, systemTablePrefixExtension, sequenceTableId,
+	)
+
+	hexCreateTableSql := getCreateTableSql(hexTableName)
+
+	// Create schema and system tables.
+	_, err = dbConn.ExecContext(ctx, `
+	CREATE SCHEMA IF NOT EXISTS public;
+
+	CREATE TABLE IF NOT EXISTS public.`+sequenceTableName+`
```

Comment:
Nice usage of the backtick string here, it makes the string statement easier to parse: 👍